src/k2/module/k23si/Module.cpp

```diff
@@ -358,6 +593,111 @@ K23SIPartitionModule::handleWrite(dto::K23SIWriteRequest&& request, dto::K23SI_M
     });
 }
 
+seastar::future<std::tuple<Status, dto::K23SIPartialUpdateResponse>>
+K23SIPartitionModule:: handlePartialUpdate(dto::K23SIPartialUpdateRequest&& request, dto::K23SI_MTR sitMTR, FastDeadline deadline) {
```

Comment:
Is it possible to reduce some of the code duplication here with the handleWrite function?

Reply:
k23si Module has been modified: 

1. handlePartialUpdate was not combined with handleWrite function. Because I think two different functions can make the code clearer to read.
2. _parsePartialRecord function was re-written. _makeFieldsForSameVersion function is added to make full-record-WI in the case of same schema and same version. _makeFieldsForDiffVersion function is added to make full-record-WI in the case of same schema and different versions.
3. unordered_map has been removed. Instead, use an vector to record the offset of each field in base SKVRecord (in function _makeFieldsForDiffVersion ).

also, added two more test cases for SKVClientTest and they are all listed below:

### Test cases for same schema version update

| test case                                                    | Expected result | Note                                                         |
| ------------------------------------------------------------ | --------------- | ------------------------------------------------------------ |
| Partial update value fields, with Skipped PartitionKey and RangeKey fields | BadParameter    | missing partition key                                        |
| Partial update value fields, including PartitionKey and RangeKey fields | KeyNotFound     | Equals to write another key. Cannot use it to create a new SKVRecord. |
| Partial update every value， using field index to indicate the fieldsToUpdate | OK              | k23si_client do not encode payload in this case. It is a quick path. |
| Partial update some value fields using field name to indicate the fieldsToUpdate | OK              |                                                              |
| Partial update some fields with NULL fieldsToUpdate          | OK              | NULL fieldsToUpdate implies copy of the previous data. All fields are not updated |
| Partial update while the sematic of excludedFields in record is different from fieldsToUpdate. <br /><br />1): fieldsToUpdate indicate a field shall be updated, but this field is skipped in the record | BadParameter    | must got a value while updating a field                      |
| 2): fieldsToUpdate indicate  a field shall not be updated, but record has a value of this field | OK              | the value in record would be dropped                         |
| Updates an null field while keeping other null fields remain null | OK              |                                                              |

### Test cases for different schema versions update

| test case                                                    | Expected result | Note |
| ------------------------------------------------------------ | --------------- | ---- |
| same fields but different orders, copy pervious data (null fieldsToUpdate) | OK              |      |
| missing a field with same field name but different field type | badParameter    |      |
| add some fields, but leaving them not updated                | badParameter    |      |
| add some fields, do not contain some of the pre-existing fields | OK              |      |
| decrease some fields, and update value-fields to null-fields | OK              |      |
| update fields from null to null                              | OK              |      |